### ISSUE TITLE
Install multiple versions of a dependency

### ISSUE THREAD

**Issue by denis-sokolov** [npm/npm#5499]
Currently a package can have only a single version of a dependency defined.
I have a need to install multiple versions of a dependency at once.
This is a subset of aliasing features that were rejected in #2943.

In my use case I have a library ([`frozen-express`](https://www.npmjs.org/package/frozen-express)) that can during runtime take an object produced by another library (Express app) and do useful things with it.
I support multiple versions of Express.

During development I want to be able to test my application with different versions of my dependency. I could fake both versions of Express, but I consider that using a real Express app is a better deal. The tests become more of integration tests as opposed to unit tests, but that only makes them more useful.

I could use `npm install express@3 && npm test && npm install express@4 && npm test`, but that is very significantly slower than running my entire test suite.
In addition, this wrecks any chance of receiving proper code coverage reports.
Obviously, I have code that conditionally applies to one or another version, and no code coverage tool understands that, if I need to run `npm test` twice (or more times).

The best workaround is, unfortunately, abusing the npm database. A single-line module [`express3`](https://github.com/denis-sokolov/express3) that depends on `express@3` and exports it solves all my problems.

Perhaps this feature request is a tiny bit more viable than #2943?

**Comment by pderaaij** [comment]
Whats the status on this issue? We are in need for such a solution so we can use multiple versions of a dependency. Our use case is the following. 

We have a dependency that contains schema object which are information contracts. Our application using these schema objects need to support multiple versions of these schemas so we can send out the correct information based on the version of the schemas used by the consumer.

So we want to define a dependency on tag V1 of schemas and tag V2 on schemas. Within the application we'll do the logic of requiring the correct version, but for now it isn't possible to define the different versions as a dependency in the package.json.

It seems to me that the proposed solution in this issue will help us out, so I'm curious what the state of this issue is.

**Comment by denis-sokolov** [comment]
Interesting use case, @pderaaij. I like it.

There's a possible cricitism: versions of your schemas should not necessarily be tied to versions of your schema package. Consider that if V1 and V2 are used in production, they're both still actual, so perhaps the versioning distinction should happen inside a single version of the schema package.

As for what the state of the issue is, the answer, I think, is the same as with most other feature requests: npm devs are flooded with work.

**Comment by medikoo** [comment]
I think this change will require major changes to node's modules resolution logic, which I think it's no go, and not really worth it.

If someone has really such rare use case, he/she should probably include needed modules versions directly in a project, and require it as e.g. `require('./lib/express/v2')`, `require('./lib/express/v3')`

@pderaaij how do you imagine require'ing different versions of same package in your module? e.g. if package is name `schema`, normally you require it with `require('schema')`. How do you see require'ing distinctively two different versions of `schema` within one module?

**Comment by pderaaij** [comment]
@denis-sokolov  Well I can't entirely agree on that, although I understand your remark. But the package containing the schemas should only contain that schemas. Not also the logic of choosing the correct version. For consumers there only should be one package containing the schemas, not more than that.

@medikoo We're thinking about using a solution like the rewire module or a dynamic requiring mechanism. We are not entirely sure yet. I see some possibilities though.

**Comment by matthew-dean** [comment]
+1. I also would love to be able to install / consume multiple versions of a package in a production app. In the case of compilers, when there are breaking changes, some users want to choose which version of a compiler they use, if they wrote older code targeting an older compiler feature set. The only way I'd see how to do that is to install packages of different versions.

**Comment by devinivy** [comment]
+1 when a peer dependency spans multiple incompatible versions of the same package, those versions need to be testable (able to be required as separate dev dependencies).

**Comment by silkentrance** [comment]
@denis-sokolov Making this feature available from only the package.json or the optional npm-shrinkwrap.json is fine, but it would be even nicer to also be able to use the command line for such a purpose.

And, looking at the existing usage for `npm install`, I find that the `@` is already in use, e.g.

```
npm install <pkg>@<tag>
npm install <pkg>@<version>
npm install <pkg>@<version range>
```

Not considering the other usages where one passes in a folder, tar ball or tar ball url or git url.

For this, we would have to provide either an `--alias=<alias name>` option or extend upon the package parameter above. And, since the `@` is already reserved for separating the version / tag from the package name, this would require something else, for example a single `:` colon?

Using the colon actually works just fine, e.g.

```
cd /tmp
mkdir test\:3
echo "module.exports.a = 1" > test\:3/index.js
node
> t3 = require('./t:3');
{ a: 1 }
```

I don't know how this behaves on the Windows platform, however, Meteor are using this sort of scheme for naming their packages, e.g. <atmosphere account name>:<package name>.

**Comment by isaacs** [comment]
I don't think that we need new fanciness for this.  It's a rare edge case, and a confusing thing to implement in a general way.

Just make a package like this:

```json
{ "name": "express3",
  "version": "1.0.0",
  "description":"Express version 3",
  "dependencies": { "express":"3" } }
```

```javascript
// index.js
module.exports = require('express')
```

**Comment by isaacs** [comment]
Note: please do not write a bot that spams the registry with a zillion packages for every major version of every module.  But if you have an actual use case for something like this, then go for it.

If someone has a dozen or a hundred *actual use cases today* for this, then let's investigate further.  But what I'm hearing so far is a lot of engineering solutions for a very small one-off problem that only a few people have, with relatively little regard for the hazards introduced by those engineering solutions.  The tradeoff just isn't profitable here, I'm afraid.

**Comment by denis-sokolov** [comment]
Thank you, @isaacs, for an extensive feedback and an official statement that a package like `express3` is not agains npm terms. This does solve the issue.

**Comment by matthew-dean** [comment]
@isaacs There are 12 people in this thread alone, 8 of them for it. Considering that not everyone who wants the feature will comment here, then it's easy to conclude that there are at least a dozen valid use cases.

**Comment by aetsher** [comment]
Tq allz☺

**Comment by silkentrance** [comment]
@isaacs good solution, however, how about preventing these zillion wrapper packages to show up on npmjs.org in the first place by indicating to these users that there exist npm caches/proxies that allow the user to publish private packages to their own registry and still be able to use the official registry?

**Comment by silkentrance** [comment]
@isaacs actually, I was working on a patch for npm and the registry... well I guess I will stop working on it then.

**Comment by charliewolf** [comment]
On the off chance numbers help, throwing another vote for multiple-version support without the need for stub packages.

**Comment by silkentrance** [comment]
Another idea, how about implementing a proxy/cache that would allow users to depend on say

```
dependencies : {
    'express:3' : ...
}
```

and which will be able to solve the problem without having to modify the existing npm code base?

**Comment by pderaaij** [comment]
This will not solve my issue and I think this is a issue worth solving. Just to easy to pass this on.

**Comment by devinivy** [comment]
In my use case I have a package that supports two major versions of hapi.  I need my tests to test both versions of hapi (so both versions need to be dev dependencies).  I solved this by cloning hapi 7 into its own Git repo.  I include the hapi 7 package in package.json using a Github URI and hapi 8 as a normal dependency.  It's really not ideal, but it works.

**Comment by silkentrance** [comment]
@devinivy this issue will only cause pollution of the npm registry if people begin to publish stub packages as pointed out by isaacs. The issue itself is about preventing such things from happening.

**Comment by unthunk** [comment]
The basic reason for wanting to include multiple versions of the same package is that there is a delta between features and that you can only get the ones you need through multiple versions.  The current example listed are all JavaScript based (Express, jQuery, hapi) and the impact on testing is mentioned multiple times.  However, with more developers using npm to manage ALL their dependencies and dev dependencies (CSS, web fonts, JavaScript, build tools, etc)  on any combo of front-end, back-end, and middleware it seems likely there will be an increasing need to support multiple versions with stubs.

**Comment by monolithed** [comment]
+1, @devinivy, I've the same situation :)

**Comment by matthew-dean** [comment]
@unthunk What you said. I think the assumption of npm is that each version is really the same product. But often between major versions, you're often talking about different products that have the same name. Some packages will even say, "if you're using X use version A, if using Y, use version B". And it's sometimes the case that you're using X and Y.

**Comment by isaacs** [comment]
So, let's break down the problems that we're really trying to solve here.  There are two that are most relevant:

1. I have a module that allegedly plays nicely with multiple versions of a framework.  I want to test against the versions that I allegedly work with.
2. I depend on a library, and need to *simultaneously* use both version X of that library, and version Y of that library, because a feature was added in Y, but in such a way that it removed or changed some other feature from version X that I depend on.

(2) is more straightforward, so I'll address that first.  This is clearly a case of poor factoring.  *Usually* the correct solution is to abstract out the functionality you're depending on.   Most frameworks simply can't work side-by-side like this anyway.  (Qv. jQuery, Hapi, Express, etc.)  No one yet has provided a real-world example of (2) in this thread, so I am somewhat suspicious that it's only hypothetical anyway.

(1) is pretty challenging, to be sure.  As someone who maintains a lot of modules that have to work with multiple versions of node, I totally get it.  However, again, you're not going to be using multiple different versions of the framework or library *at the same time*, 

The alternate solution is to have a test script that sets up a few different working dirs, and installs the different versions of the framework in multiple places.  Or, create the shim module for the framework/version in question.  There's probably less than 12 shim modules required.  That's fine.

It's important to also look at the cost in terms of [broken symmetry](https://github.com/npm/newww/issues/576#issuecomment-71520032) here: the simplest and most dependable part of the mental model of npm packages would no longer be a guarantee.  It's no longer the case that the `"name"` field is what's on the website, the key in the `dependencies` hash, the folder on disk, etc.  If we implement something like this, we will be adding a new dimension to that field, and **permanently** complicating the model.

It may seem like a trivial esthetic thing, but it's quite an important decision, and not one we could ever un-make.  It's not something we should do lightly, or put to a popular vote of 12 people in a github issue.  I need you all to understand this.  Breaking that symmetry is a "last resort, break in case of emergency" kind of option, to be considered once all possible options have been exhausted.

None of the use cases presented here have been explored in anywhere near enough detail to justify making that call.

If I've missed some of the subtleties of the use cases involved here, then please let me know.

I don't see any of this being particularly tied to "front end" vs "back end" concerns.

**Comment by mzgol** [comment]
@isaacs Thanks for outlining your concerns. I understand the fear of introducing such additional mental complexity and since you stated that in those situations where there's a need of testing on different major versions of a library it's acceptable to create wrapper packages, it's enough for me.

**Comment by analog-nico** [comment]
@isaacs +1

I explained my jQuery use case above which IMHO is a valid one. Since I used Bower in this case I don't need this feature in npm. Browserify users may have this use case though. But even then this occurs for only a few libraries I guess. Those could be published with a "legacy" wrapper package.

It would be great if everyone else who suggested a use case should explain why npm should solve that. For some it should be the build script's job instead.

FWIW I am not against that feature but also not pro anymore.

**Comment by monolithed** [comment]
@analog-nico, I absolutely agree with you. The Bower is a crutch/correction for npm (

**Comment by isaacs** [comment]
@analog-nico Your jQuery case is my (1) case above.  It can be handled with a shim module very easily.  You'll never need both jQuery 1.x and 2.x in production at the same time (in fact, that wouldn't even work).

**Comment by unthunk** [comment]
@isaacs I think your statement makes the case that npm is opinionated into how packages are/should be used/consumed by projects.  Sure, (2) could very well be a sign of poor factoring in many cases.  But in the case of html imports and web components, it's completely feasible that a project may want to include multiple versions of the same package.  To use jQuery as a contrived example, an author of a web component may wish to build two versions of the same component from within a project - one built with 2.x to be consumed by Cordova or other web based mobile frameworks and another with 1.x to be used on web pages.  More generally, because html imports scope HTML, CSS, and JS within a particular component, it should be up to the author of the project to decide is there is a valid reason for imports included in a project to use different versions of a package (and not to nitpick, but we are talking about packages, otherwise it would be called node library manager :P).

**Comment by silkentrance** [comment]
I believe that @isaacs is right in that for different, provably incompatible, versions of a given package, one must actually be writing different test cases.

As such, setting up a test project for each of these different versions might make more sense than to have the top level project pull different versions for a given package.

Eg.

```
root/
   test/
      test-for-version3/
          package.json
          test/
       test-for-version4/
           package.json
           test/
```

This can also be applied for versions that are actually API compatible but still must be tested independently.

**Comment by isaacs** [comment]
@unthunk 

> npm is opinionated into how packages are/should be used/consumed by projects.

To a certain extent, yes, npm *is* a bit opinionated!  It's a lot less opinionated than some package managers, and allows for a lot of customization through the package.json and configuration settings, but without *some* opinions, it would just be wget.

One opinion that npm has is that the key in the `dependencies` object, the `name` field, and the argument to `require()` are all the same.  This isn't the first time that this idea was brought up.  The suggested feature is essentially the same as the old `mappings` proposal for the CommonJS Packages specification.  All my arguments from that debate apply equally here, and the simplicity that comes from npm's opinion is worth having to work around the special cases when something like this is truly necessary.

The fact that Web Components scope JS and CSS means that two *different* web components might use different versions of JQuery.  npm supports this.  (In fact, it is one of the *only* package managers that supports this!)  However, it does not support having two different things both named "JQuery" at the same node in the dependency graph.  Adding that support means breaking a very valuable bit of symmetry.  It is a very costly change, and not one that I'm going to allow for a rare use case with clear workarounds; even less so for a hypothetical use-case that I don't think anyone actually has.

I've been thinking it'd be interesting to write a test runner that simplifies the work of running your tests against multiple different versions of a dependency.  It could maybe fiddle with the `NODE_PATH` environ and run the tests in child procs, or load them in some special way that monkeypatches `require()` like Mocha and Lab do.  I think that would be a much better solution to the real problems brought up in this thread.

**Comment by unthunk** [comment]
> but without some opinions, it would just be wget

@isaacs that is a gross oversimplification and not quite true.  It's fine for a package manager to have opinions or not - as always, pick the right tool for the job.  

> The fact that Web Components scope JS and CSS means that two different web components might use different versions of JQuery. npm supports this

Yes, when a project is consuming different web components that are registered as separate packages then the dependencies are in separate trees. But what if the web components are part of the project itself (possibly not a best practice) and are not packages?  And what about authors that may wish to provide builds for jQuery 1.x and 2.x, or Fontawesome 3.x and 4.x, or any other lib, all within the same package?
   
Something to consider is that people considering using npm as their sole package manager may be using Bower which is self-described as unopinionated.  Bower makes it very simple to install a package using a different name `<name>=<package>#<version>` and people may be looking for parity with this in npm.  Understandably, npm would want to keep parity with package names and require() given that npm was built with expectation that node would be consuming packages.  So in the case the someone has a need (for whatever reason) to include two versions of the same package in their project, they should either use a stub as mentioned above if they want to use npm or mange those particular dependencies outside of npm.

**Comment by shawnExplore** [comment]
I happen to have at least a dozen use cases. 

_Backgrounder:_
We use npm modules in NetSuite (a Rhino based server-side javascript solution). It is common here to deploy a _single_ script file to perform some business logic and that script can reference libraries. Each script and its libraries are separate at runtime.

_Our use case:_
Over some months or even years we may deploy 6 or 10 or 20 scripts to production.  Oldest scripts may use version 1.0 of a library. Newer scripts may use 1.5, and brand new scripts may want 2.0. All these scripts are deployed in production and each is _able to reference libraries independently_. In this model new scripts can use 2.0 without changing old scripts already using an earlier version. All the scripts are still part of the overall "project".

I believe the workaround presented here (uniquely naming versions of a package) will work for us, but wanted to throw in another use case. In our situation we have the scenario above duplicated across dozens of customers.

**Comment by othiym23** [comment]
@shawnExplore If you squint a little, those scripts you describe can be treated as plugins, and in the case of plugins, it makes sense to give each one its own install root. If there's shared functionality that can be used across all of them, you can either explicitly install those dependencies higher up the tree, or express the dependency through `peerDependencies` (although as of `npm@3` this will become a distinction without a difference, because you'll already need to install `peerDependencies` explicitly). Once you have these scripts installed in their own subdirectories, with their own independently varying sets of dependencies, you no longer to have multiple versions of  a single dependency in a single place in the dependency tree. Is there something I'm missing here?

**Comment by tkoomzaaskz** [comment]
Hello everyone, I couldn't read entire thread, but I found you were giving usecases for multiple npm package version. So here is mine: I'm using browserify to build a package of my library for an online demo. This online demo is bound to a specific version of the library, of course. It'd be nice if I could browserify multiple versions and include them in the demo using a dropdown select, for instance. So far I could only do this manually and break entire automation process.

**Comment by skylize** [comment]
I want to add another use case. I'll use lodash as the most obvious example because it's such a common dependency. I have 3.9.3 installed, but my dependencies require older versions. I have 5 copies installed just for the archaic version 1.0.2. I should be able to install it at the top of my tree for them to share, but I can't because the top of my tree already has the new version. Lodash itself is pretty small, but I have so many copies because it's required by larger packages that suffer the same issue.

I propose something like `npm install --save-common lodash@1` or 
`npm install --save-common-dev lodash@1`. 

Npm would at that point:
1. Install the module into a folder with the version number appended to the name
2. Replace all matching modules in the dependency tree with symlinks to the versioned module. 
3. Mark it in as a common dependency in package.json using token characters, so ``npm install`` would be able to replicate the process.

This would also provide reasonable workarounds for the other use cases described, because the versioned module could be required using it's folder name.

**Comment by baslr** [comment]
Hi,
I would like to serve statically multiple versions of eg. `angular`
```
Projects that depend on v1.4.1
https://static.x.y/node_modules/angular@1.4.1/angular.js

Projects that depend on v1.3.16
https://static.x.y/node_modules/angular@1.3.16/angular.js

Projects that depend on v1.2.28
https://static.x.y/node_modules/angular@1.2.28/angular.js
```

I'm like oO that multiple version installation is not possible.

bower does `bower install jquery-legacy=jquery#1.10 jquery-modern=jquery#2`

Maybe:
```
npm install --save angular@1.4.1 as angular141
; node_modules/angular141/
npm install --save angular@1.3.16 as angular1316
; node_modules/angular1316/
```

**Comment by monolithed** [comment]
@baslr, interesting idea :+1:

**Comment by silkentrance** [comment]
@monolithed @baslr Personally, I think that the solution npm install -g express:3@^3.0.0 is best because it allows you install multiple packages using the same command line. And it also works just as fine in the package descriptor.

See https://github.com/npm/npm/issues/5499#issuecomment-69495168

**Comment by dlong500** [comment]
Here's a use case for multiple versions of jQuery in production (via webpack): using jQuery 1.x for client-side code (to support IE8) and jQuery 2.x for server-side node.js code.  Not that I'm using jQuery on the server side at the moment, but since only jQuery 2.x can be used with node that seems like a valid use case for needing both of them in production.  There can certainly be hacks to get around it, but still...

**Comment by leo** [comment]
Hey folks!

@serapath pointed me here.

We're currently [discussing][1] a different method of managing modules and our suggestion also includes a way to keep multiple versions of a single module on a machine within a global directory. And I think that's also solving your problem.

And I would be very pleased if you could also say something about this (of course you don't need to entirely agree with us, just show us want you think about this or how we could improve this whole idea).

Thanks in advance! :blush:

[1]: https://github.com/nodejs/NG/issues/20

**Comment by matthew-dean** [comment]
@Leo That's great news, but for me, it's not globally where I need multiple versions, but locally (in a project folder).

**Comment by leo** [comment]
@matthew-dean What's the specific reason why you need them installed locally?

**Comment by zhanzhenzhen** [comment]
Git URL as Dependency already gives us this feature:
```json
{
    "project-v1": "git://github.com/user/project#1.0.0",
    "project-v2": "git://github.com/user/project#2.0.0"
}
```
We can easily give each a name, and thus use the two versions simultaneously because they are treated as two packages.
But while using npm packages we can't rename it. It is like that we already have a feature but when our project is hosted on the official npm repository this feature is taken away. I think the syntax should be expanded to have the same power as URL as dependency.

Example:
```json
{
    "project-v1": "project@1.0.0",
    "project-v2": "project@2.0.0",
}
```

So we can:
```javascript
var project1 = require("project-v1");
var project2 = require("project-v2");
```

I also would like this feature to be implemented both globally and locally (at least just for consistency).

**Comment by denis-sokolov** [comment]
> Git URL as Dependency already gives us this feature

I have to admit I did not think of that. A neat workaround!

**Comment by aredridel** [comment]
I've also created alias packages: `foo-2` just exports `foo@2`'s exports, and `foo-1` exports `foo@1`'s exports

**Comment by silkentrance** [comment]
+1 for zhanzhenzhen´s suggestion

**Comment by devinivy** [comment]
I also created an alias package, but I recognize it's a worrisome pattern... it makes sense to keep npm clean.

**Comment by zhanzhenzhen** [comment]
I tested... and found I've made a mistake. The name on the left doesn't have any effect for Git URLs. It still uses the package's "name" field in its package.json. If two packages with the same name are listed, only the first will be installed.

**Comment by denis-sokolov** [comment]
Oh, okay. At least you can fork and change the `name` field without submitting it to npm's registry.

**Comment by serapath** [comment]
Maybe it would make sense to collect all use cases somewhere to really figure out if that feature is necessary. I guess stats from npm, how often people actually create aliases for certain packages would be interesting to know too

**Comment by tenbits** [comment]
My use case is:
We have created benchmark tests for our library to test the performance of different versions of the library. As for now, we use `bower` and `git releases`, for instance
```javascript
{
  "dependencies": {
    "foo-0.52.4": "https://github.com/xxx/foo.git#0.52.4",
    "foo-0.12.23": "https://github.com/xxx/foo.git#0.12.23"
  }
}
```
Requiring: `require('./bower_components/foo-VERSION/lib/mask.node.js')`

It is absolutely not critical, but it would be nice to have at least similar behaviour with `npm`, while as for now it is not possible to have different git tags of the same package in `package.json`

**Comment by zhanzhenzhen** [comment]
@isaacs 

>So, let's break down the problems that we're really trying to solve here. There are two that are most relevant:

>1. I have a module that allegedly plays nicely with multiple versions of a framework. I want to test against the versions that I allegedly work with.
>2. I depend on a library, and need to simultaneously use both version X of that library, and version Y of that library, because a feature was added in Y, but in such a way that it removed or changed some other feature from version X that I depend on.

I think there's a 3rd use case: merge two projects into one. Project A is using one version of a library, while project B is using another incompatible version of that library. Though later the merged code can be totally rewritten to use only one version, it may take time, so if you want it to run just after merge, you will need this feature.

**Comment by denis-sokolov** [comment]
@zhanzhenzhen, I think your proposed 3rd case can be reduced to the second use case, though.

**Comment by cztomsik** [comment]
It is at least possible to write/use other dependency manager? Npm is great as a repository but not so good as a tool.

This is is really something npm **should support right from the start**. I'm not sure if you're doing any other projects than npm itself but features like these are very handy in real-world (sub-optimal) web applications...

**Comment by mnpenner** [comment]
My million+ line app now has 9 dependencies that are a full major version or two behind. There's no way I can just upgrade them without everything breaking. 

Now that we've implemented webpack, it would be very easy for use to `require` different versions of the dependencies in different parts of the app, but there's no easy way for us to actually *install* the different versions concurrently. A feature like this would be very useful to us.

**Comment by silkentrance** [comment]
When closing such an important issue, please feel free to at least provide a comment :grin:

**Comment by othiym23** [comment]
@silkentrance There were two comments by @isaacs around the closing of this issue, so I'm not sure to what you're referring.

**Comment by silkentrance** [comment]
@othiym23 To my excuse, it was late and I did not see that it was already closed in 2015 :dancer:

**Comment by matthew-dean** [comment]
@othiym23 Which comment are you referring to? The first one, where @isaacs said that use cases were needed to see why this should happen, or the second, where he said that no matter how many use cases people had, this wouldn't happen?

**Comment by othiym23** [comment]
The second. This isn't an aspect of npm's design that is going to change.

**Comment by cztomsik** [comment]
Rewriting the whole project to a new version of library at once is hard - it may be even **so hard it will get rejected by project lead** and then people may get stuck with legacy code just like in Java.

I think this wasn't an issue before because there were no long-existing projects - node was just a toy platform but this has changed - this should be reconsidered or somebody should give us a hint **how to migrate to new versions without doing it all at once**.

**Comment by isiahmeadows** [comment]
@cztomsik If you can use scoped packages, you could just do several local packages with different versions to ease migration, and delete the old ones once everything has been migrated. It is a little bit of an abuse, but it works.

**Comment by cztomsik** [comment]
@isiahmeadows good to know, thx!

**Comment by mjsalinger** [comment]
@isaacs My own use case for your example (2)

I have an Express API which I'm currently running using a specific package behind the scenes, a particular middleware that is core to the package.  The middleware releases a new SemVer-major release which is a major breaking API change that also introduces some new features that are very relevant.  

I want to preserve my existing stability in the API by introducing a new API version that makes use of the new API version, refactors some things, and adds the new features.  I don't want to rewrite my API-current version to use the new API used by the middleware.

**Comment by jbrumwell** [comment]
+1 for me

**Comment by trusktr** [comment]
I would love this. My use case is #2 in theone's @isaacs pointed out:


> I depend on a library, and need to simultaneously use both version X of that library, and version Y of that library, because a feature was added in Y, but in such a way that it removed or changed some other feature from version X that I depend on.

I am using a 3D rendering engine to make demos, and each demo is made with a different version of that engine. Compiling each demo statically with webpack is a pain, and I also don't want to go and update each demo to use the latest API changes of the 3D engine.

It'd be amazing to import different version of the same package for each demo.

This is very easy to do with JSPM, for example. Wish it was the case with NPM.

**Comment by trusktr** [comment]
A possible workaround to this is making each part of code that depends on different version of the same package modules themselves, with their respective dependencies on different version of the same package. In this case, `npm install` will place the differing versions of the dependency in each node_modules folder of the package-ized pieces of code. This will work.

**Comment by isiahmeadows** [comment]
I have a proposed fix for this: allow dependencies to be installed under aliases. What about something like this:

```json
"dependencies": {
  "express as express-2": "2",
  "express as express-3": "3",
  "express as express-4": "4"
}
```

What do you all think of this?

Furthermore, I see this as potentially useful in four other cases:

1. You're migrating from Underscore to Lodash, but you use a few Underscore-specific features. You can insert `"lodash as underscore": "4.x.x"` in your `package.json`, and write a compatibility wrapper while you migrate, that you later remove.
2. You use a popular client-side framework through Browserify called `framework` that has limited support for Node, but you wish to use it fully isomorphically. You need full support, even down to the data model. So, you decide to re-implement the API with the help of a few others within your company, and use that instead. You still want to use the existing module name for it for aesthetic reasons, though, so you want to alias it like `"@company/framework-node as framework": "0.1.*"`.
3. You use a framework with an annoyingly long name that you don't want to type for half your files. For purely ergonomic reasons, you and your other project members want to use a much shorter name, just to spare your fingers from repetitive strain injury. Sadly, management doesn't want to change the name of the scoped package itself, and isn't budging. So, you alias this scoped package locally to `"@courier-mail-supply/javascript-browser-framework as framework": "2.*"`.
4. You saw a gist, which had an `index.js`, but no `package.json`, and wish to use that in your project. You add it via git hash, and alias it to something that's a little more descriptive than the hexadecimal ID that it comes with.

-----

I will note an important caveat with this approach: when you choose an alias, that effectively becomes the local package name. So, given this `package.json`:

```json
{"devDependencies": {"to-list as listify": "*"}}
```

`npm` will error early if you try running `npm install listify`. If you want to uninstall that module, either `npm uninstall to-list` or `npm uninstall listify` will work, since they both refer to the same local package. And given the below `package.json`,

```json
"dependencies": {
  "express as express-2": "2",
  "express as express-3": "3",
  "express as express-4": "4"
}
```

- `npm uninstall express` will uninstall all three packages.
- `npm uninstall express-2` or `npm uninstall express@2` will uninstall the first entry.
- `npm uninstall express-3` or `npm uninstall express@3` will uninstall the second entry.
- `npm uninstall express-4` or `npm uninstall express@4` will uninstall the third entry.

Furthermore, `npm` will warn on every command with this `package.json` until it gets fixed, and `npm install`, etc. will error early:

```json
{
  "devDependencies": {
    "@myorg/to-list as listify": "*",
    "listify": "1.5.3"
  }
}
```

**Comment by scott113341** [comment]
Hi all, I stumbled upon this thread while developing a benchmarking tool.  I want to be able to install multiple versions of a package and make them "compete" against each other.  

[`npm-install-version`](https://github.com/scott113341/npm-install-version) is what I came up with.  Essentially, you can do things like this: 

```javascript
const niv = require('npm-install-version');

niv.install('csjs@1.0.0');
// installs csjs@1.0.0 to node_modules/csjs@1.0.0/

niv.install('csjs@1.0.1');
// installs csjs@1.0.1 to node_modules/csjs@1.0.1/

niv.install('scott113341/csjs#some-branch');
// installs https://github.com/scott113341/csjs#some-branch to node_modules/scott113341-csjs#some-branch/

const csjs_old = niv.require('csjs@1.0.0');
const csjs_new = niv.require('csjs@1.0.1');
const csjs_branch = niv.require('scott113341/csjs#some-branch');
```

It's not an ideal solution, but it works.  You can also use it via CLI.  [Feedback welcome!](https://github.com/scott113341/npm-install-version/issues)

[Here](https://github.com/scott113341/csjs-benchmarks/blob/master/index.js) is how I'm using it to run benchmarks.

**Comment by le0nik** [comment]
Is there any way to do this now?
I need to install 2 `webpack` versions: `1.13.1 and `2.1.0-beta.20` for integration testing.

So far I've tried:

- Github urls. 2nd overwrites the first one, so `webpack-1` doesn't exist(probably has to do with `npm@3` flat `node_modules`)
```json
{
    "webpack-1": "git://github.com/webpack/webpack#v1.13.1",
    "webpack-2": "git://github.com/webpack/webpack#v2.1.0-beta.20",
}
```
- `npm-install-version`. Doesn't work at all. Install webpack, but fails on require. `Error: Cannot find module 'webpack@1'`. `directrory` option doesn't work at all.

- `node-multidep`(https://github.com/joliss/node-multidep). Doesn't work either. `Error: Cannot find module 'tap'`

Does anyone know any other ways?

**Comment by denis-sokolov** [comment]
@le0nik, [a stub module](https://github.com/npm/npm/issues/5499#issuecomment-71089419).

**Comment by le0nik** [comment]
@denis-sokolov, thanks! Was hoping for a better solution, but it really seems like this is the only working one.

**Comment by scott113341** [comment]
Hey @le0nik, if you have a moment would you mind opening an [issue](https://github.com/scott113341/npm-install-version/issues) for the `npm-install-version` problem you're having?  Would be much appreciated!

**Comment by le0nik** [comment]
@scott113341 sure, no problem, will open an issue with detailed report asap.

**Comment by ainthek** [comment]
stub/proxy packages as suggested by @isaacs are fine for modules that has a single export,
and are only accesses by require("module"), however there are several modules (good or bad) that
are used as require("module/xyz") and then version require("module1/xyz") will not work,
so probably the proxying solution is to write module.exports=require(); and maybe add module.exports.require("path", version) as final pattern ?

**Comment by isiahmeadows** [comment]
That sounds like a Node extension, which IMHO is a bad idea, considering
this is just a package management issue.

**Comment by protometa** [comment]
@le0nik, this might also be solved by refactoring the different parts of your module into separate modules that rely on different versions of some module.

E.g., `my-webpack-module` might require `my-webpack-module-v1-tests` and `my-webpack-module-v2-tests` which depend on `webpack@1.13.1` and `webpack#v2.1.0-beta.20` respectively.

**Comment by adamcee** [comment]
My solution to this was to install using the github link. npm allows you to specifiy the branch installed.
http://stackoverflow.com/questions/16350673/depend-on-a-branch-or-tag-using-a-git-url-in-a-package-json
In my case, I added "material-ui-next" to my package.json, followed by the appropriate link.

Note that npm install actually installed the "next" branch of material ui to node_modules/material-ui-build-next

**Comment by cztomsik** [comment]
@adamcee good but it might not work for all packages - sometimes there's a build step before publishing to npm and those files would have been missing...

**Comment by adamcee** [comment]
Good point. And, I actually ended up using a wrapper package which someone
else had published.

**Comment by TimBeyer** [comment]
This could also be nice for API versioning.  
I mostly structure my services in a way where the HTTP API is more or less a small adapter to a library that actually does the hard work.  
If I could install multiple versions of the same package, I'd be able to switch between two or more handlers for the different versions of the API, each of them importing and using the library version that corresponds to that API version.  
That would make it much easier to maintain and still fix older API versions and have a clear separation without any `if(version === 2) doSomethingElse()`.

**Comment by tommysullivan** [comment]
If you polled all programmers and asked how many of them needed more than one version of a package in the same codebase in the last 20 years you would most certainly find thousands.

Today we have an API that abstracts away the details of using ElasticSearch. It uses an elasticsearch client package version X that works with ES server version X. Users want to now be able to use our API against ES server version Y. Client version X is incompatible with ES Server Y; however Client version Y is compatible with ES server Y. I expected to be able to solve this by depending on both versions of the client and then choosing the implementation at runtime based on discovery of the ES server version, a very reasonable expectation in my opinion.

I will take advantage of the shim workaround, but I do hope that (a) this type of case is recognized as a completely legitimate programming endeavor and (b) a better solution is eventually identified and implemented in a way that minimizes the negative implications mentioned in earlier comments.

**Comment by nyurik** [comment]
I have ran into this exact problem in Wikipedia.  The graph subsystem uses vega library. Every Vega version is not backward compatible. Yet, I need to be able to support all of the legacy Vega versions, because user generated content relies on that version's syntax. So I had to create and maintain my own separate Vega forks, one per version. https://github.com/wikimedia/mediawiki-services-graphoid/blob/master/package.json#L45

Something like  `"dependencies": { "vega#1": "vega#1, "vega#2": "vega#2", ... }`, and corresponding `require('vega#1')` would have been much easier and less error prone to maintain.

**Comment by fny** [comment]
I'm running into this problem. I have a Webpack plugin that I need to test against all Webpack versions. Right now the only way I can do that is by reinstalling each package on the run for each version.

**Comment by nyurik** [comment]
@isaacs please reconsider and re-open this bug - there is clearly a community need for this feature as indicated by numerous comments. The lack of a clear solution does not mean the issues is not there - but simply that there is no clear path forward yet.  Keeping it open will indicate dev's willingness to understand the need, and will allow everyone to constructively provide a good solution (or possibly even a PR)

**Comment by victorherraiz** [comment]
if this is reopened at some point it could be great to have all the alternatives in one place with the pros and cons to avoid going in circles

**Comment by vorillaz** [comment]
This is a common issue for my team as well. We are currently migrating a huge web app (~150 LOC) built with Backbone Marionette. Since the major releases bring lots of breaking changes we want to break down our app and create independent bundles with different Marionette versions in order to progressively refactor the whole project. Thus, we are stuck, since there is no chance to use NPM as a package manager. We finally used NPM and a local repository in order to manage our dependencies.

**Comment by adamlubek** [comment]
another use case is that I have angular app which is composed of smaller feature npm packages. I want to be able to load specific version of package depending on client e.g. all clients would be presented with package version 1.0.33 but one of the clients would be presented with 1.1.0. This way I could verify new features only on one client before rolling out to all clients.

**Comment by hinell** [comment]
I agree that the best way to manage multiple modules is aliasing. Installation path could be changed by just specifying new folder name on the left-side of its dependency field as following: 
* package.json#dependecies
```js
{
  "foo"     : ">=2.0.0" ,
  "foo-old" : "foo@1.9.0"
}
```
Where: 
* ``foo`` is installed as usually into folder named as foo: ``node_modules/foo/``;
equivalent to ``"foo":"foo@>=2.0.0" ``
* ``foo-old`` is installed into ``node_modules/foo-old/``
The loading of a modules could be as following:

```js
var foo    = require('foo') // getting  2.0.0 version
var fooold = require('foo-old') // 1.9.0
```

I think it can be implemented pretty easily.

**Comment by victorherraiz** [comment]
I really like the aliasing idea. Another implementation option that could be simpler (not better) to build:

``` json
{
  "aliases": {
    "xreq-old": "xreq"
  },
  "dependencies": {
    "xreq": "^2.0.0",
    "xreq-old": "^1.0.0",
    }
  }
}
```
`aliases` section could create local alias.

**Comment by TomMahle** [comment]
I wanted to post here for anyone like me that is using Yarn and landed here. It is a more or less drop-in replacement for NPM that supports aliasing out of the box:

yarn add material-ui@latest
yarn add material-ui-next@npm:material-ui@next
then

import FlatButton from 'material-ui/FlatButton'; // v0.x
import Button from 'material-ui-next/Button'; // v1.x

(credit for example goes to https://github.com/callemall/material-ui/issues/7195#issuecomment-314547601 )

**Comment by chandlervdw** [comment]
And for those that want to mostly use **material-ui-next**:
```
yarn add material-ui-old@npm:material-ui@latest
yarn add material-ui@next
```
then
```
import Button from 'material-ui-old/DatePicker'; // v0.x
import { pink } from 'material-ui/colors'; // v1.x
```

**Comment by nyurik** [comment]
Running `yarn add material-ui-old@npm:material-ui@latest` adds `"material-ui-old": "npm:material-ui@latest"` line to `package.json`'s `dependencies` section, but at least npm 5.5.1 throws an error for `npm install`.  So while amazing, using this feature means npm can no longer be used for the project.

**Comment by matthew-dean** [comment]
@nyurik I think the point is to not use NPM at that point, if you're going to have aliased dependencies. Yarn is becoming fairly popular for NPM management these days.

**Comment by nyurik** [comment]
@matthew-dean of course, but many projects explicitly allow users to use both packaging systems interchangeably, and this split won't be very healthy for the community. Especially in cases where there is a large number of different projects but a common build system, and some projects require a different manager.  I sincerely hope NPM team would reconsider their decision.

**Comment by nemanjacosovic** [comment]
Why is this still even a thing? Its very clear that we need multiple versions. I don't see any valid reasons against it.

**Comment by skylize** [comment]
@nemanjacosovic This issue has only been racking up comments with pleas to implement and ideas on how to do it since June 2014? Why should we have expected any movement on this yet? Rofl

Actually you will find it's been closed since January 2015. So I guess that's when the maintainers just quit listening to us.

**Comment by silkentrance** [comment]
@skylize if so, they could have shut down comments on this thread a while ago, but they did not.

**Comment by scott113341** [comment]
@skylize your assertion that the maintainers of npm don't listen to its community of users is, frankly, disrespectful.  There are dozens of comments from maintainers in this issue alone explaining very well why this functionality may be very difficult to implement properly.

If you think that there is a particular solution (either discussed above or not yet explored) that is robust enough to overcome these well-explained technical barriers, then I would kindly ask that you contribute to the technical discussion by explaining why the solution is adequate.

**Comment by skylize** [comment]
@scott113341 
> your assertion that the maintainers of npm don't listen to its community of users is, frankly, disrespectful.

I'm greatly appreciative of all the work that's gone into `npm`, so I apologize to anyone who might feel disrespected by my comments. 

Unfortunately, I guess I let my real feelings spill out about this particular issue. The claim that you should only ever need one version of a dependency seems absurdly near-sighted to me in today's world of deeply nested graphs of rapidly updating dependencies with breaking changes.

I remember reading this post way back when  [from @isaacs on Jan 22, 2015](https://github.com/npm/npm/issues/5499#issuecomment-71089419), which permanently closed the issue:
> I don't think that we need new fanciness for this. It's a rare edge case, and a confusing thing to implement in a general way.

This post did not raise any real technical concerns or explain what is so difficult. It was just a raw, unfiltered opinion from one person that shut everything down for everybody.  From what I can tell, this seems to be not-even-on-the-radar for the people actually writing code for the project, because it is a "closed" issue. Put this on a roadmap and we might see some real progress and some practical solutions.

In a [later comment](https://github.com/npm/npm/issues/5499#issuecomment-71089419) @isaacs adds

>If someone has a dozen or a hundred actual use cases today for this, then let's investigate further. But what I'm hearing so far is a lot of engineering solutions for a very small one-off problem that only a few people have.

The community has filled this thread to the brim with numerous use-cases, but the issue remains closed and seemingly ignored.

---
> If you think that there is a particular solution (either discussed above or not yet explored) that is robust enough to overcome these well-explained technical barriers, then I would kindly ask that you contribute to the technical discussion by explaining why the solution is adequate.

I added [a proposal in June 2015](https://github.com/npm/npm/issues/5499#issuecomment-111204293) on how to deal with the overall technical problems, which nobody ever raised any concerns against.

> I propose something like `npm install --save-common lodash@1` or `npm install --save-common-dev lodash@1`.
> Npm would at that point:
> 1. Install the module into a folder with the version number appended to the name
> 2. Replace all matching modules in the dependency tree with symlinks to the versioned module.
> 3. Mark it in as a common dependency in package.json using token characters, so npm install would be able to replicate the process.

Looking back at that I see some flaws in my idea, namely the reliance on symlinks which are not easily dealt with cross-platform. But platform-specific symlinks are not impossible. And regardless, the basic premise of writing packages to version-numbered folders is still a good starting point, if someone was actually trying to build this.

**Comment by mnpenner** [comment]
Just wanted to remind you guys that `yarn add lodash2@npm:lodash@2.x` *already works* so we *know* this is possible, and there's already a precedent for the syntax for it. It looks like this in `package.json`:

```json
  "dependencies": {
    "lodash2": "npm:lodash@2.x"
  },
```

I think this is a better solution than the one proposed by @skylize too because it lets you choose a name for the additional packages, opposed to just appending a number to the name which may or may not conflict with another package. Also, I don't think there's a need for the special cli-args, `--save-common`.

**Comment by skylize** [comment]
@mnpenner 

> I think this is a better solution than the one proposed by @skylize too because it lets you choose a name for the additional packages

In this [comment to #2943](https://github.com/npm/npm/issues/2943#issuecomment-10388316), @isaacs made it pretty clear that blanket aliasing will _never_ be considered. If we want a solution that will actually get built, we need to be thinking of ways to give the user only the bare minimum amount of control needed to reach the target, not allowing them to just name anything whatever they want.

**Comment by mnpenner** [comment]
@skylize I think what isaacs was getting at is that is if you `require` one module, it shouldn't load the module from an unrelated directory (I think). With yarn's solution, `lodash2` is actually installed into `node_modules/lodash2`. So node's module resolution algorithm remains completely unchanged.

That said, yes, you could use this deceive the reader by aliasing one package to an unrelated package's name, but I don't understand why you would want to go to such great lengths to prevent developers from writing stupid code. I could just as easily clone `jquery` and publish to npm as `lodash2` and then what? A few people get confused and the world chugs on.

**Comment by nyurik** [comment]
@skylize I don't think "never consider" is ever a good approach when working with a community. Opinions change, new arguments and usecases come forth, etc.  The #2943 comment was made over 5 years ago - yarn didn't even exist back then. In just a year, yarn is a major player in this field - clearly showing that NPM has become the IE6 of the package management, and could use some love.

`yarn` clearly broke compatibility on this one. Yet, the community is clearly in favor of `yarn`'s approach, so I think NPM team should reconsider. Otherwise it hurts all of us - we all would like both, the compatibility AND the version aliasing. If NPM introduces a new incompatible syntax just for version aliasing without any justification, the issue will get even worse.

**Comment by dead-claudia** [comment]
Could we start trying sending support emails or something, rather than blowing this issue up with borderline off-topic stuff? It kind of gets old getting an email + a notification each time one of you all comment here.

**Comment by DuBistKomisch** [comment]
In response to https://github.com/npm/npm/issues/5499#issuecomment-71930827:

>No one yet has provided a real-world example of (2) in this thread, so I am somewhat suspicious that it's only hypothetical anyway.

We want to use `socket.io-client` to connect to separate v1 and v2 servers, and the protocols are mutually incompatible, so we'd need both clients installed simultaneously.

Eventually I'd like to switch to yarn for this project of course (sick of npm's shit, and yarn's only burned me once so far), but I don't have time to clean up all the tech debt required right now, so this is really annoying... guess I'll try out `npm-install-version`/`node-multidep`.

Edit: `npm-install-version` works I guess, can't actually include it in the `package.json` though so just need to install it manually everywhere...

Edit 2: `npm isntall -g yarn` gg ez

**Comment by palmerj3** [comment]
I have a real-world scenario for this.. I maintain jest-junit and I would like to setup travis ci to test it against multiple versions of jest. Doing that is fine but I would also like to be able to run my unit tests with whatever version of jest I choose.

**Comment by hinell** [comment]
@palmerj3 Well, there are a lot of real-world scenarios especially when you writing cross-platform software but looks like npm is unlikely to implement anything cause the issue is opened almost 3 years ago and nothing has changed since then. That's why it makes me sad. 😞

**Comment by silkentrance** [comment]
While it is fun seeing that this is not going to be resolved, whereas for example yarn provides a very good alternative, I am now signing off.

**Comment by trevorwhealy** [comment]
👋 We have a lot of legacy code written in the v1 release of React-Transition-Group (https://github.com/reactjs/react-transition-group), and it is working great. 

As React evolved and the maintainers changed, the codebase also changed dramatically, causing them to release a v2 that had breaking changes. It would be a lot of work to retool legacy code that is actually working well and won't be changed for the foreseeable future.

We locked down to v1 so that our app would still work, but we want to use v2 for all future development. Allowing aliases would be pretty useful in this case. 

This comment from earlier in the thread seems to be a simple solution to this common problem: https://github.com/npm/npm/issues/5499#issuecomment-322761229

Ideally, we would really enjoy being able to alias like so:
```js
"react-transition-group@v1": "^1.1.3",
"react-transition-group@v2": "https://github.com/reactjs/react-transition-group.git#master",
"react-transition-group@v3: "^3.0.0",
```

**Comment by jlpospisil** [comment]
I am working on migrating a large project from bootstrap 3 to 4.  This feature would be nice.  I would like to import both, and apply bootstrap 4 to specific pages/areas as the migration progresses.

**Comment by pauldraper** [comment]
Yarn supports this case via general-purpose [aliasing](https://twitter.com/sebmck/status/873958247304232961). ("This isn't going to be reconsidered" #2943)`

```
"react-transition-group-v1": "npm:react-transition-group@^1.1.3",
"react-transition-group-v2": "https://github.com/reactjs/react-transition-group.git#master",
"react-transition-group-v3: "npm:react-transition-group@^3.0.0",
```

**Comment by VincentCharpentier** [comment]
So basically is this gonna be a feature ?
With the constant (breaking) changes we have to face in javascript projects this seems like an obviously needed feature.
Will it be possible soon ?
If not, that's not a big issue I could switch to yarn, but honestly I wish I could stay with npm
Thanks

**Comment by jamesmfriedman** [comment]
Same issue. I run a UI library for React and I need to test across 6 different versions since the react api is subtlety different between them.

**Comment by armanriazi** [comment]
According to the following links how match package.json for angular 5 and 6 
[](https://github.com/scott113341/npm-install-version/issues/28)
In my use case I need to have both of them on "Dependency" section.It's not important how handle with every form.

key : "npm-install-version"
`{ "name": "app-angular", "version": "1.0.0", "scripts": { "npm-install-version": "npm-install-version", "ng": "ng", "start": "ng serve", "build": "ng build", "test": "ng test", "lint": "ng lint", "e2e": "ng e2e" }, "private": true, "dependencies": { "core-js": "^2.5.4", "zone.js": "^0.8.26" }, "devDependencies": { "@angular-devkit/build-angular": "~0.6.1", "@angular/cli": "~6.0.1", "@angular/compiler-cli": "^6.0.0", "@angular/language-service": "^6.0.0", "@types/jasmine": "~2.8.6", "@types/jasminewd2": "~2.0.3", "@types/node": "~8.9.4", "codelyzer": "~4.2.1", "jasmine-core": "~2.99.1", "jasmine-spec-reporter": "~4.2.1", "karma": "~1.7.1", "karma-chrome-launcher": "~2.2.0", "karma-coverage-istanbul-reporter": "~1.4.2", "karma-jasmine": "~1.1.1", "karma-jasmine-html-reporter": "^0.2.2", "npm-install-version": "^6.0.2", "protractor": "~5.3.0", "ts-node": "~5.0.1", "tslint": "~5.9.1", "typescript": "~2.7.2" }, "npm-install-version": { "@angular/animations": [ "^6.0.0", "^5.1.0" ], "@angular/common": [ "^6.0.0", "^5.1.0" ], "@angular/compiler": [ "^6.0.0", "^5.1.0" ], "@angular/core": [ "^6.0.0", "^5.1.0" ], "@angular/forms": [ "^6.0.0", "^5.1.0" ], "@angular/http": [ "^6.0.0", "^5.1.0" ], "@angular/platform-browser": [ "^6.0.0", "^5.1.0" ], "@angular/platform-browser-dynamic": [ "^6.0.0", "^5.1.0" ], "@angular/router": [ "^6.0.0", "^5.1.0" ], "rxjs": [ "^6.0.0", "^5.5.2" ] } }`

**Comment by armanriazi** [comment]
Because i need Angular 5 for GraphQl and Angular 6 for stable material UI !!?

**Comment by Mani-RRI** [comment]
hello every one,
I need both rxjs@^5.5.0 and rxjs@^6.0.0-beta.4 have to be installed. But if I try to install one of these, the other gets vanish. Do I have a solution?

**Comment by nyurik** [comment]
@Mani-RRI yes, we switched to `yarn`, and used their aliasing system.  Works well.

**Comment by Mani-RRI** [comment]
@nyurik , thank you : )

**Comment by doesdev** [comment]
I really don't want to switch to Yarn. I hate that such a key part of the ecosystem is fragmenting. That being said, this thread is extremely discouraging when considering staying with NPM. 
:worried:
The team has ignored all the clearly laid out and highly valid reasons why this is needed and pretended that this is a non-issue. If it were a non-issue it would, quite literally, not be an issue. It is, and with an eff ton off comments and support from the community for said feature.

**Comment by Mani-RRI** [comment]
@doesdev  there is an easy solution, but it depends on you which versions you want to use. There are some particular sets of dependencies **work fine together**. And you need to install them together. After instaling check your JSON file to ensure that all the dependencies are available that you wanted to be installed. for example, one set of working dependencies I am providing here,... 

![json angular](https://user-images.githubusercontent.com/19514852/41019349-d7a05f68-697b-11e8-9ecb-bc8f8b3463fd.png)

**Comment by nyurik** [comment]
@Mani-RRI the whole idea of package manager is that you don't want to know what works well together. It should "just work" (TM).  Even if two versions only differs in the readme, or if a totally new code was swapped in, I as a developer should not be aware of that. I should be able to alias two or more versions, and use them side by side.
